File: src/harness_core/dispatch/observability.py
"""Phase 6A: Observability — structured logging, metrics, request tracing."""

from __future__ import annotations

import json
import logging
import threading
import time
import uuid
from collections import deque
from dataclasses import asdict, dataclass, field
from typing import Any


OBSERVABILITY_SCHEMA_VERSION = "observability.v1"


@dataclass(frozen=True)
class RequestMetric:
    request_id: str
    component: str
    action: str
    duration_ms: float
    status: str
    timestamp: float


@dataclass(frozen=True)
class MetricSnapshot:
    name: str
    value: float
    labels: dict[str, str] = field(default_factory=dict)
    timestamp: float = field(default_factory=time.time)


@dataclass(frozen=True)
class SpanRecord:
    trace_id: str
    span_id: str
    parent_span_id: str | None
    name: str
    start_time: float
    end_time: float | None
    status: str


class StructuredFormatter(logging.Formatter):
    """JSON log formatter for structured logging."""

    def format(self, record: logging.LogRecord) -> str:
        log_entry: dict[str, Any] = {
            "timestamp": record.created,
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
        }
        if hasattr(record, "trace_id"):
            log_entry["trace_id"] = record.trace_id  # type: ignore[attr-defined]
        if hasattr(record, "span_id"):
            log_entry["span_id"] = record.span_id  # type: ignore[attr-defined]
        if record.exc_info and record.exc_info[0] is not None:
            log_entry["exception"] = self.formatException(record.exc_info)
        return json.dumps(log_entry, default=str)


class MetricsCollector:
    """Ring-buffer metrics collector with thread-safe recording."""

    def __init__(self, max_size: int = 1000) -> None:
        self._max_size = max_size
        self._metrics: deque[RequestMetric] = deque(maxlen=max_size)
        self._snapshots: deque[MetricSnapshot] = deque(maxlen=max_size)
        self._lock = threading.Lock()

    def record(self, metric: RequestMetric) -> None:
        with self._lock:
            self._metrics.append(metric)

    def record_snapshot(self, snapshot: MetricSnapshot) -> None:
        with self._lock:
            self._snapshots.append(snapshot)

    def query(self, component: str | None = None, action: str | None = None) -> list[RequestMetric]:
        with self._lock:
            result = list(self._metrics)
        if component is not None:
            result = [m for m in result if m.component == component]
        if action is not None:
            result = [m for m in result if m.action == action]
        return result

    def query_snapshots(self, name: str | None = None) -> list[MetricSnapshot]:
        with self._lock:
            result = list(self._snapshots)
        if name is not None:
            result = [s for s in result if s.name == name]
        return result

    def count(self, component: str | None = None) -> int:
        return len(self.query(component=component))

    def clear(self) -> None:
        with self._lock:
            self._metrics.clear()
            self._snapshots.clear()


class RequestTracer:
    """Trace/span propagation for request lifecycle tracking."""

    def __init__(self) -> None:
        self._spans: dict[str, SpanRecord] = {}
        self._lock = threading.Lock()

    def new_trace_id(self) -> str:
        return uuid.uuid4().hex

    def new_span_id(self) -> str:
        return uuid.uuid4().hex[:16]

    def start_span(self, name: str, trace_id: str | None = None,
                   parent_span_id: str | None = None) -> tuple[str, str]:
        if trace_id is None:
            trace_id = self.new_trace_id()
        span_id = self.new_span_id()
        span = SpanRecord(
            trace_id=trace_id,
            span_id=span_id,
            parent_span_id=parent_span_id,
            name=name,
            start_time=time.time(),
            end_time=None,
            status="in_progress",
        )
        with self._lock:
            self._spans[span_id] = span
        return trace_id, span_id

    def end_span(self, span_id: str, status: str = "ok") -> SpanRecord | None:
        with self._lock:
            span = self._spans.get(span_id)
            if span is None:
                return None
            span = SpanRecord(
                trace_id=span.trace_id,
                span_id=span.span_id,
                parent_span_id=span.parent_span_id,
                name=span.name,
                start_time=span.start_time,
                end_time=time.time(),
                status=status,
            )
            self._spans[span_id] = span
        return span

    def get_span(self, span_id: str) -> SpanRecord | None:
        with self._lock:
            return self._spans.get(span_id)

    def get_trace_spans(self, trace_id: str) -> list[SpanRecord]:
        with self._lock:
            return [s for s in self._spans.values() if s.trace_id == trace_id]

    def clear(self) -> None:
        with self._lock:
            self._spans.clear()


def setup_structured_logging(
    logger_name: str = "harness_core",
    level: int = logging.INFO,
    handler: logging.Handler | None = None,
) -> logging.Logger:
    """Configure structured JSON logging for the harness_core namespace."""
    logger = logging.getLogger(logger_name)
    logger.setLevel(level)
    if handler is None:
        handler = logging.StreamHandler()
    handler.setFormatter(StructuredFormatter())
    if not any(isinstance(h.formatter, StructuredFormatter) for h in logger.handlers):
        logger.addHandler(handler)
    return logger

File: engine/src/infrastructure/observability.rs
use serde::{Deserialize, Serialize};
use std::collections::VecDeque;
use std::sync::Mutex;

pub const OBSERVABILITY_SCHEMA_VERSION: &str = "observability.v1";

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RequestMetric {
    pub request_id: String,
    pub component: String,
    pub action: String,
    pub duration_ms: f64,
    pub status: String,
    pub timestamp: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MetricSnapshot {
    pub name: String,
    pub value: f64,
    pub labels: std::collections::HashMap<String, String>,
    pub timestamp: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SpanRecord {
    pub trace_id: String,
    pub span_id: String,
    pub parent_span_id: Option<String>,
    pub name: String,
    pub start_time: f64,
    pub end_time: Option<f64>,
    pub status: String,
}

pub struct MetricsCollector {
    max_size: usize,
    metrics: Mutex<VecDeque<RequestMetric>>,
    snapshots: Mutex<VecDeque<MetricSnapshot>>,
}

impl MetricsCollector {
    pub fn new(max_size: usize) -> Self {
        Self {
            max_size,
            metrics: Mutex::new(VecDeque::with_capacity(max_size.min(1000))),
            snapshots: Mutex::new(VecDeque::with_capacity(max_size.min(1000))),
        }
    }

    pub fn record(&self, metric: RequestMetric) {
        let mut metrics = self.metrics.lock().unwrap();
        if metrics.len() >= self.max_size {
            metrics.pop_front();
        }
        metrics.push_back(metric);
    }

    pub fn record_snapshot(&self, snapshot: MetricSnapshot) {
        let mut snapshots = self.snapshots.lock().unwrap();
        if snapshots.len() >= self.max_size {
            snapshots.pop_front();
        }
        snapshots.push_back(snapshot);
    }

    pub fn query(&self, component: Option<&str>, action: Option<&str>) -> Vec<RequestMetric> {
        let metrics = self.metrics.lock().unwrap();
        metrics
            .iter()
            .filter(|m| component.is_none_or(|c| m.component == c))
            .filter(|m| action.is_none_or(|a| m.action == a))
            .cloned()
            .collect()
    }

    pub fn query_snapshots(&self, name: Option<&str>) -> Vec<MetricSnapshot> {
        let snapshots = self.snapshots.lock().unwrap();
        snapshots
            .iter()
            .filter(|s| name.is_none_or(|n| s.name == n))
            .cloned()
            .collect()
    }

    pub fn count(&self, component: Option<&str>) -> usize {
        self.query(component, None).len()
    }

    pub fn clear(&self) {
        self.metrics.lock().unwrap().clear();
        self.snapshots.lock().unwrap().clear();
    }
}

pub struct RequestTracer {
    spans: Mutex<std::collections::HashMap<String, SpanRecord>>,
    counter: Mutex<u64>,
}

impl Default for RequestTracer {
    fn default() -> Self {
        Self::new()
    }
}

impl RequestTracer {
    pub fn new() -> Self {
        Self {
            spans: Mutex::new(std::collections::HashMap::new()),
            counter: Mutex::new(0),
        }
    }

    pub fn new_trace_id(&self) -> String {
        let mut counter = self.counter.lock().unwrap();
        *counter += 1;
        let c = *counter;
        format!("trace-{c:08x}")
    }

    pub fn new_span_id(&self) -> String {
        let mut counter = self.counter.lock().unwrap();
        *counter += 1;
        let c = *counter;
        format!("span-{c:08x}")
    }

    pub fn start_span(
        &self,
        name: &str,
        trace_id: Option<&str>,
        parent_span_id: Option<&str>,
        now: f64,
    ) -> (String, String) {
        let trace_id = trace_id
            .map(String::from)
            .unwrap_or_else(|| self.new_trace_id());
        let span_id = self.new_span_id();
        let span = SpanRecord {
            trace_id: trace_id.clone(),
            span_id: span_id.clone(),
            parent_span_id: parent_span_id.map(String::from),
            name: name.to_string(),
            start_time: now,
            end_time: None,
            status: "in_progress".to_string(),
        };
        self.spans.lock().unwrap().insert(span_id.clone(), span);
        (trace_id, span_id)
    }

    pub fn end_span(&self, span_id: &str, status: &str, now: f64) -> Option<SpanRecord> {
        let mut spans = self.spans.lock().unwrap();
        let span = spans.get(span_id)?.clone();
        let updated = SpanRecord {
            end_time: Some(now),
            status: status.to_string(),
            ..span
        };
        spans.insert(span_id.to_string(), updated.clone());
        Some(updated)
    }

    pub fn get_span(&self, span_id: &str) -> Option<SpanRecord> {
        self.spans.lock().unwrap().get(span_id).cloned()
    }

    pub fn get_trace_spans(&self, trace_id: &str) -> Vec<SpanRecord> {
        self.spans
            .lock()
            .unwrap()
            .values()
            .filter(|s| s.trace_id == trace_id)
            .cloned()
            .collect()
    }

    pub fn clear(&self) {
        self.spans.lock().unwrap().clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn metrics_collector_records_and_queries() {
        let collector = MetricsCollector::new(100);
        collector.record(RequestMetric {
            request_id: "r1".into(),
            component: "http".into(),
            action: "GET /health".into(),
            duration_ms: 1.5,
            status: "200".into(),
            timestamp: 1000.0,
        });
        collector.record(RequestMetric {
            request_id: "r2".into(),
            component: "scheduler".into(),
            action: "tick".into(),
            duration_ms: 5.0,
            status: "error".into(),
            timestamp: 1001.0,
        });
        assert_eq!(collector.count(None), 2);
        assert_eq!(collector.count(Some("http")), 1);
        assert_eq!(collector.count(Some("scheduler")), 1);
        let errors = collector.query(Some("scheduler"), None);
        assert_eq!(errors[0].status, "error");
    }

    #[test]
    fn metrics_collector_bounded_ring() {
        let collector = MetricsCollector::new(2);
        for i in 0..5 {
            collector.record(RequestMetric {
                request_id: format!("r{}", i),
                component: "test".into(),
                action: "a".into(),
                duration_ms: 0.0,
                status: "ok".into(),
                timestamp: i as f64,
            });
        }
        assert_eq!(collector.count(None), 2);
        let metrics = collector.query(None, None);
        assert_eq!(metrics[0].request_id, "r3");
        assert_eq!(metrics[1].request_id, "r4");
    }

    #[test]
    fn tracer_spans_lifecycle() {
        let tracer = RequestTracer::new();
        let (trace_id, span_id) = tracer.start_span("request", None, None, 100.0);
        assert!(trace_id.starts_with("trace-"));
        assert!(span_id.starts_with("span-"));

        let span = tracer.get_span(&span_id).unwrap();
        assert_eq!(span.status, "in_progress");
        assert!(span.end_time.is_none());

        let ended = tracer.end_span(&span_id, "ok", 105.0).unwrap();
        assert_eq!(ended.status, "ok");
        assert_eq!(ended.end_time, Some(105.0));

        let trace_spans = tracer.get_trace_spans(&trace_id);
        assert_eq!(trace_spans.len(), 1);
    }

    #[test]
    fn snapshot_recording() {
        let collector = MetricsCollector::new(100);
        collector.record_snapshot(MetricSnapshot {
            name: "scheduler.tick".into(),
            value: 42.0,
            labels: [("executor".into(), "noop".into())].into(),
            timestamp: 1000.0,
        });
        let snaps = collector.query_snapshots(None);
        assert_eq!(snaps.len(), 1);
        assert_eq!(snaps[0].name, "scheduler.tick");
        assert_eq!(snaps[0].value, 42.0);
    }
}
